docs/proposals/high-availability.md

```diff
@@ -0,0 +1,34 @@
+
+## Design Options
+For complete reference see [this](https://www.ibm.com/developerworks/community/blogs/RohitShetty/entry/high_availability_cold_warm_hot?lang=en)
+
+1. Hot Standby: In this scenario, data and state are shared between the two components such that an immediate failure in one component causes the the standby deamon to take over exactly where the failed component had left off.  This would be an ideal solution for kubernetes, however it poses a series of challenges in the case of controllers where component-state is cached locally and not persisted in a transactional way to a storage facility.  This would also introduce additional load on the apiserver, which is not desireable.  As a result, we are **NOT** planning on this approach at this time. 
+
+2. **Warm Standby**: In this scenario there is only one active component acting as the master and additional components running by not providing service or responding to requests.  Data and state are not shared between the active and standby components.  When a failure occurs, the standby component that becomes the master must determine the current state of the system before resuming functionality.
+
+3. Active-Active (Load Balanced): Clients can simply load-balance across any number of servers that are currently running.  Their general availability can be continuously updated, or published, such that load balancing only occurs across active participants.  This aspect of HA is outside of the scope of *this* proposal because there is already a partial implementation in the apiserver.
+
+## Design Discussion Notes on Leader Election
+Implementation References:
+* [zookeeper](http://zookeeper.apache.org/doc/trunk/recipes.html#sc_leaderElection)
+* [etcd](https://groups.google.com/forum/#!topic/etcd-dev/EbAa4fjypb4)
+* [initialPOC](https://github.com/rrati/etcd-ha)
+
+In HA, the apiserver will provide an api for sets of replicated clients to do master election: acquire the lease, renew the lease, and release the lease.  This api is component agnostic, so a client will need to provide the component type and the lease duration when attemping to become master.  The lease duration should be tuned per component.  The apiserver will attempt to create a key in etcd based on the component type that contains the client's hostname/ip and port information. This key will be created with a ttl from the lease duration provided in the request.  Failure to create this key means there is already a master of that component type, and the error from etcd will propigate to the client.  Successfully creating the key means the client making the request is the master.  Only the current master can renew the lease.  When renewing the lease, the apiserver will update the existing key with a new ttl.  The location in etcd for the HA keys is TBD.
```

Comment:
Since we are now proposing a generic HA master election API (open to users and not restricted to infrastructure components), I believe we need to ensure that an unrelated client/component cannot take over the lease when the elected master fails. One way to go about this is to use namespaces for writing these keys. Then the administrator can write its keys for the infrastructure components into its own namespace (default ?) and clients can do so in their own namespaces.

Very crudely, something on the lines of:
```
.../<namespace>/leader-keys/<component-type1>
.../<namespace>/leader-keys/<component-type2>
```
This still assumes that multiple "types" of components within a namespace have to play nicely with each other, but that is still a much more manageable problem.